src/librustc_typeck/check/coercion.rs

```diff
@@ -447,6 +503,48 @@ impl<'f, 'gcx, 'tcx> Coerce<'f, 'gcx, 'tcx> {
         success(adjustments, ty, obligations)
     }
 
+    // T -> impl Bounds, where T satisfies Bounds
+    fn coerce_hidden(&self, source: Ty<'tcx>, target: Ty<'tcx>) -> CoerceResult<'tcx> {
+        debug!("coerce_hidden(source={:?}, target={:?})", source, target);
+
+        let target_predicates = if let ty::Opaque(def_id, substs) = target.sty {
```

Comment:
Hmm, I'm not sure if this is where I think this code should go or not. It depends on whether we want to support e.g. `(impl Foo, impl Bar)` etc. 

One way to go about this would be to look to see whether the type includes any `ty::Opaque` **for which this location is a defining use**. I don't remember off the top of my head what the code here looks like, @oli-obk may be able to point us at the right place, else I can dig harder.

But we may want to -- for now, at least -- move this out of the "main coercion" path and instead up to the code that handles a `let x: T = ...` statement. We would in that case look to see whether `T` contains any `impl Trait` statements, much as we do for a function return: if so, we would use `instantiate_opaque_types` to create a type with fresh variables, and then coerce from the source type into that type. Then we'd take the user's type annotation and return *that* as the type for the underlying pattern. Something like that.

Reply:
I feel like this is very analogous to normalization. Ultimately, we'll probably want to be handling it in a "lazy fashion" somehow. Have to think about what that means I guess.

Reply:
The think that makes me nervous here:

Can this coercion code trigger at times when it shouldn't?

In particular, opaque types play this "dual role", and I'm worried we'll confuse it here.

Example:

```rust
fn foo() -> impl Debug { }

fn main() {
    let x = if true {
        foo() // return type will be a TyOpaque
    } else {
        22 // I think this can "coerce" to the type from the `then` arm
    };
}
```

Here, I imagine we should get an error, but we might wind up going down *this* path and trying to add a `Hide` annotation or something?